Developed function to updateApplication by adding a version

yashodgayashan opened the pull request.

yashodgayashan commented:
Constructing application is using in two database function and I am thinking of make a utility function to construct the application using TreeReamovalForm.
What do you think about it? @deshankoswatte @sanjulamadurapperuma @vikumkbv

sanjulamadurapperuma commented on core/src/env_platform/db_handler.bal:
```diff
@@ -155,7 +156,68 @@ function updateApplicationDraft(TreeRemovalForm form, string applicationId) retu
+    log:printDebug("Constructed application: " + application.toString());
+
+    // Add the new version to the versions array
+    versions.push(application);
```
Since the only real difference between **updateApplicationDraft** method and **updateApplication** is this line and the versions array definition, I suggest removing this newly added **updateApplication** function and moving this line to the updateApplicationDraft method. We could pass in a **boolean value** to figure out whether the update _is for making a new version of the application or not_. 

@yashodgayashan @deshankoswatte @vikumkbv let me know if there is a reason if this change is not possible.

yashodgayashan replied:
Yeah it can be done. Since update application draft will alter the existing version but update application will get the versions array and push the new version to the versions array and update the application document. Shall we make a one function ?

deshankoswatte replied:
+1 for @sanjulamadurapperuma's comment.

deshankoswatte replied:
+1 to @sanjulamadurapperuma's comment

sanjulamadurapperuma commented on core/src/env_platform/db_handler.bal:
```diff
@@ -155,7 +156,68 @@ function updateApplicationDraft(TreeRemovalForm form, string applicationId) retu
+    log:printDebug("The application of application id: " + applicationId.toString() + " is " + found.toString());
+
+    // Get the versions array.
+    json[] versions = <json[]>found[0].versions;
```
This can also be moved to the updateApplicationDraft function.

yashodgayashan commented:
@sanjulamadurapperuma this might be the one you suggesting.

```
function updateApplicationDraft(TreeRemovalForm form, string applicationId, string applicationType) returns boolean|error {

    map<json> application = {
        "title": form.title,
        "applicationCreatedDate": {
            "year": form.applicationCreatedDate.year,
            "month": form.applicationCreatedDate.month,
            "day": form.applicationCreatedDate.day,
            "hour": form.applicationCreatedDate.hour,
            "minute": form.applicationCreatedDate.minute
        },
        "removalDate": {
            "year": form.removalDate.year,
            "month": form.removalDate.month,
            "day": form.removalDate.day,
            "hour": form.removalDate.hour,
            "minute": form.removalDate.minute
        },
        "reason": form.reason,
        "applicationType": form.applicationType,
        "requestedBy": form.requestedBy,
        "permitRequired": form.permitRequired,
        "landOwner": form.landOwner,
        "treeRemovalAuthority": form.treeRemovalAuthority,
        "city": form.city,
        "district": form.district,
        "nameOfTheLand": form.nameOfTheLand,
        "planNumber": form.planNumber,
        "area": extractAreaAsJSONArray(form.area),
        "treeInformation": extractTreeInformationAsJSONArray(form.treeInformation)
    };
    log:printDebug("Constructed application: " + application.toString());

    int|mongodb:DatabaseError updated ;

    if(applicationType == "draft"){
        updated = applicationCollection->update({"versions.0": application}, {"applicationId": applicationId});
    } else {
        // Get the exsisting application.
        map<json>[] found = check applicationCollection->find({"applicationId": applicationId});
        log:printDebug("The application of application id: " + applicationId.toString() + " is " + found.toString());

        // Get the versions array.
        json[] versions = <json[]>found[0].versions;
        versions.push(application);

        // Added new versions array to the application
        updated = applicationCollection->update({"versions": versions}, {"applicationId": applicationId});
    }

    if (updated is int) {
        log:printDebug("Updated status for application with application ID: " + applicationId + " is " + updated.toString() + ".");
        return updated == 1 ? true : false;
    } else {
        log:printDebug("An error occurred while updating the draft application with the application ID: " + applicationId + ". " + updated.reason().toString() + ".");
        return updated;
    }
}
```

yashodgayashan commented:
```
# The `updateApplicationDraft` function will alter the exsisting application draft with the incoming form details.
# 
# + form - Form containing the tree removal data.
# + applicationId - The Id of the application which should be altered.
# + return - This function will return true if draft is updated in the database, false if not or else it returns a mongodb:Database error.
function updateApplicationDraft(TreeRemovalForm form, string applicationId) returns boolean|error {

    map<json> application = constructApplication(form);
    log:printDebug("Constructed application: " + application.toString());

    int|mongodb:DatabaseError updated = applicationCollection->update({"versions.0": application}, {"applicationId": applicationId});

    if (updated is int) {
        log:printDebug("Updated status for application with application ID: " + applicationId + " is " + updated.toString() + ".");
        return updated == 1 ? true : false;
    } else {
        log:printDebug("An error occurred while updating the draft application with the application ID: " + applicationId + ". " + updated.reason().toString() + ".");
        return updated;
    }
}

# The `updateApplication` function will add a new version of an application with the incoming form details.
# 
# + form - Form containing the tree removal data.
# + applicationId - The Id of the application which should be added.
# + return - This function will return true if version of the application is updated in the database, 
# false if not or else it returns a mongodb:Database error.
function updateApplication(TreeRemovalForm form, string applicationId) returns boolean|error {

    // Get the exsisting application.
    map<json>[] found = check applicationCollection->find({"applicationId": applicationId});
    log:printDebug("The application of application id: " + applicationId.toString() + " is " + found.toString());

    // Get the versions array.
    json[] versions = <json[]>found[0].versions;

    map<json> application = constructApplication(form);
    log:printDebug("Constructed application: " + application.toString());

    // Add the new version to the versions array
    versions.push(application);

    // Added new versions array to the application
    int|mongodb:DatabaseError updated = applicationCollection->update({"versions": versions}, {"applicationId": applicationId});

    if (updated is int) {
        log:printDebug("Updated status for application with application ID: " + applicationId + " is " + updated.toString() + ".");
        return updated == 1 ? true : false;
    } else {
        log:printDebug("An error occurred while updating the draft application with the application ID: " + applicationId + ". " + updated.reason().toString() + ".");
        return updated;
    }
}
```

@sanjulamadurapperuma @deshankoswatte This is the way I suggest where we make a utility function to construct the application. What is the best one?

sanjulamadurapperuma commented:
> ```
> # The `updateApplicationDraft` function will alter the exsisting application draft with the incoming form details.
> # 
> # + form - Form containing the tree removal data.
> # + applicationId - The Id of the application which should be altered.
> # + return - This function will return true if draft is updated in the database, false if not or else it returns a mongodb:Database error.
> function updateApplicationDraft(TreeRemovalForm form, string applicationId) returns boolean|error {
> 
>     map<json> application = constructApplication(form);
>     log:printDebug("Constructed application: " + application.toString());
> 
>     int|mongodb:DatabaseError updated = applicationCollection->update({"versions.0": application}, {"applicationId": applicationId});
> 
>     if (updated is int) {
>         log:printDebug("Updated status for application with application ID: " + applicationId + " is " + updated.toString() + ".");
>         return updated == 1 ? true : false;
>     } else {
>         log:printDebug("An error occurred while updating the draft application with the application ID: " + applicationId + ". " + updated.reason().toString() + ".");
>         return updated;
>     }
> }
> 
> # The `updateApplication` function will add a new version of an application with the incoming form details.
> # 
> # + form - Form containing the tree removal data.
> # + applicationId - The Id of the application which should be added.
> # + return - This function will return true if version of the application is updated in the database, 
> # false if not or else it returns a mongodb:Database error.
> function updateApplication(TreeRemovalForm form, string applicationId) returns boolean|error {
> 
>     // Get the exsisting application.
>     map<json>[] found = check applicationCollection->find({"applicationId": applicationId});
>     log:printDebug("The application of application id: " + applicationId.toString() + " is " + found.toString());
> 
>     // Get the versions array.
>     json[] versions = <json[]>found[0].versions;
> 
>     map<json> application = constructApplication(form);
>     log:printDebug("Constructed application: " + application.toString());
> 
>     // Add the new version to the versions array
>     versions.push(application);
> 
>     // Added new versions array to the application
>     int|mongodb:DatabaseError updated = applicationCollection->update({"versions": versions}, {"applicationId": applicationId});
> 
>     if (updated is int) {
>         log:printDebug("Updated status for application with application ID: " + applicationId + " is " + updated.toString() + ".");
>         return updated == 1 ? true : false;
>     } else {
>         log:printDebug("An error occurred while updating the draft application with the application ID: " + applicationId + ". " + updated.reason().toString() + ".");
>         return updated;
>     }
> }
> ```
> 
> @sanjulamadurapperuma @deshankoswatte This is the way I suggest where we make a utility function to construct the application. What is the best one?

Yes a utility function to construct the application would be ideal as it would be used across a lot of functions

sanjulamadurapperuma commented:
> @sanjulamadurapperuma this might be the one you suggesting.
> 
> ```
> function updateApplicationDraft(TreeRemovalForm form, string applicationId, string applicationType) returns boolean|error {
> 
>     map<json> application = {
>         "title": form.title,
>         "applicationCreatedDate": {
>             "year": form.applicationCreatedDate.year,
>             "month": form.applicationCreatedDate.month,
>             "day": form.applicationCreatedDate.day,
>             "hour": form.applicationCreatedDate.hour,
>             "minute": form.applicationCreatedDate.minute
>         },
>         "removalDate": {
>             "year": form.removalDate.year,
>             "month": form.removalDate.month,
>             "day": form.removalDate.day,
>             "hour": form.removalDate.hour,
>             "minute": form.removalDate.minute
>         },
>         "reason": form.reason,
>         "applicationType": form.applicationType,
>         "requestedBy": form.requestedBy,
>         "permitRequired": form.permitRequired,
>         "landOwner": form.landOwner,
>         "treeRemovalAuthority": form.treeRemovalAuthority,
>         "city": form.city,
>         "district": form.district,
>         "nameOfTheLand": form.nameOfTheLand,
>         "planNumber": form.planNumber,
>         "area": extractAreaAsJSONArray(form.area),
>         "treeInformation": extractTreeInformationAsJSONArray(form.treeInformation)
>     };
>     log:printDebug("Constructed application: " + application.toString());
> 
>     int|mongodb:DatabaseError updated ;
> 
>     if(applicationType == "draft"){
>         updated = applicationCollection->update({"versions.0": application}, {"applicationId": applicationId});
>     } else {
>         // Get the exsisting application.
>         map<json>[] found = check applicationCollection->find({"applicationId": applicationId});
>         log:printDebug("The application of application id: " + applicationId.toString() + " is " + found.toString());
> 
>         // Get the versions array.
>         json[] versions = <json[]>found[0].versions;
>         versions.push(application);
> 
>         // Added new versions array to the application
>         updated = applicationCollection->update({"versions": versions}, {"applicationId": applicationId});
>     }
> 
>     if (updated is int) {
>         log:printDebug("Updated status for application with application ID: " + applicationId + " is " + updated.toString() + ".");
>         return updated == 1 ? true : false;
>     } else {
>         log:printDebug("An error occurred while updating the draft application with the application ID: " + applicationId + ". " + updated.reason().toString() + ".");
>         return updated;
>     }
> }
> ```

Yes, it would help reduce code duplication

yashodgayashan commented:
```
function updateApplication(TreeRemovalForm form, string applicationId, string applicationType) returns boolean|error {

    map<json> application = constructApplication(form);
    log:printDebug("Constructed application: " + application.toString());

    int|mongodb:DatabaseError updated ;

    if(applicationType == "draft"){
        updated = applicationCollection->update({"versions.0": application}, {"applicationId": applicationId});
    } else {
        // Get the exsisting application.
        map<json>[] found = check applicationCollection->find({"applicationId": applicationId});
        log:printDebug("The application of application id: " + applicationId.toString() + " is " + found.toString());

        // Get the versions array.
        json[] versions = <json[]>found[0].versions;
        versions.push(application);

        // Added new versions array to the application
        updated = applicationCollection->update({"versions": versions}, {"applicationId": applicationId});
    }

    if (updated is int) {
        log:printDebug("Updated status for application with application ID: " + applicationId + " is " + updated.toString() + ".");
        return updated == 1 ? true : false;
    } else {
        log:printDebug("An error occurred while updating the draft application with the application ID: " + applicationId + ". " + updated.reason().toString() + ".");
        return updated;
    }
}
```

@sanjulamadurapperuma @deshankoswatte I think this might be the simplest one which will make a function to constrict the application map<json> as suggested by @sanjulamadurapperuma

sanjulamadurapperuma commented:
> ```
> function updateApplication(TreeRemovalForm form, string applicationId, string applicationType) returns boolean|error {
> 
>     map<json> application = constructApplication(form);
>     log:printDebug("Constructed application: " + application.toString());
> 
>     int|mongodb:DatabaseError updated ;
> 
>     if(applicationType == "draft"){
>         updated = applicationCollection->update({"versions.0": application}, {"applicationId": applicationId});
>     } else {
>         // Get the exsisting application.
>         map<json>[] found = check applicationCollection->find({"applicationId": applicationId});
>         log:printDebug("The application of application id: " + applicationId.toString() + " is " + found.toString());
> 
>         // Get the versions array.
>         json[] versions = <json[]>found[0].versions;
>         versions.push(application);
> 
>         // Added new versions array to the application
>         updated = applicationCollection->update({"versions": versions}, {"applicationId": applicationId});
>     }
> 
>     if (updated is int) {
>         log:printDebug("Updated status for application with application ID: " + applicationId + " is " + updated.toString() + ".");
>         return updated == 1 ? true : false;
>     } else {
>         log:printDebug("An error occurred while updating the draft application with the application ID: " + applicationId + ". " + updated.reason().toString() + ".");
>         return updated;
>     }
> }
> ```
> 
> @sanjulamadurapperuma @deshankoswatte I think this might be the simplest one which will make a function to constrict the application map as suggested by @sanjulamadurapperuma

+1

deshankoswatte commented on core/src/env_platform/db_handler.bal:
```diff
@@ -62,9 +62,9 @@ function saveApplication(TreeRemovalForm form) returns boolean|error {
 
     if (inserted is mongodb:DatabaseError) {
-        log:printDebug("An error occurred while saving the application with ID: " + application.applicationId + ". " + inserted.reason() + ".") ;
+        log:printDebug("An error occurred while saving the application with ID: " + application.applicationId.toString() + ". " + inserted.reason().toString() + ".");
```
Isn't the reason in string format?

yashodgayashan replied:
Yes.

deshankoswatte replied:
Then why are we calling the `toString()` there?

yashodgayashan replied:
I will check it.

sanjulamadurapperuma replied:
Yes I think we can remove that

yashodgayashan replied:
<img width="1433" alt="Screenshot 2020-05-16 at 08 07 07" src="https://user-images.githubusercontent.com/32588172/82108546-c6646b80-974c-11ea-8624-8d218af51347.png">

sanjulamadurapperuma replied:
Yes @yashodgayashan is right, we have to toString the reason, otherwise it throws an error. This is because of the complications when doing string concatenations in Ballerina I believe.

yashodgayashan replied:
@sanjulamadurapperuma Yes. Then we have to go with toString()

yashodgayashan commented:
@deshankoswatte @sanjulamadurapperuma Resolved requested changes and please specially go through line number 130 - 152

sanjulamadurapperuma commented on core/src/env_platform/db_handler.bal:
```diff
@@ -112,50 +112,50 @@ function getApplicationStatusByApplicationId(string applicationId) returns strin
+    // If the exsiting application is a draft.
+    if (applicationStatus == "draft") {
+        if (form.applicationType == "draft") {
+            updated = applicationCollection->update({"versions.0": application}, {"applicationId": applicationId});
```
since this updates the 0th index in versions array, does it mean that we are overriding whatever application is already in that index?

yashodgayashan replied:
Yes. We don't keep draft application versions. Then it is good to overriding the existing one. But when it comes to the submitted ones we have to make versions.

sanjulamadurapperuma replied:
That part is yet to be done right?

yashodgayashan replied:
It did't get it. The overall logic is 
* If the application with the given application ID is draft and incoming form is also a draft then, it will alter and add to the versions[0]. 
* If the application with the given application ID is draft and incoming form is submit then, it will alter and add to the versions[0] and change the status to draft.
* If the application with the giver application ID is submitted and the incoming form is submitted then, it will add the application to the versions array.
* else give appropriate errors.

sanjulamadurapperuma replied:
Right okay. My concern was whether we need to manage versions for applicatonStatus = draft and applicationType = submit. I guess that's not needed then.

yashodgayashan replied:
No it is not needed.

yashodgayashan commented:
@sanjulamadurapperuma @deshankoswatte Any concerns?

deshankoswatte commented on core/src/env_platform/db_handler.bal:
```diff
@@ -112,50 +112,50 @@ function getApplicationStatusByApplicationId(string applicationId) returns strin
 }
 
-# The `updateApplicationDraft` function will alter the exsisting application draft with the incoming form details.
+# The `updateApplication` function will either alter the exsisting application draft with the incoming form details or 
+# add a new version of an application with the incoming form details.
```
```suggestion
# The `updateApplication` function will either alter the existing application draft or add a new version for an application 
# with the incoming form details.
```

deshankoswatte commented on core/src/env_platform/db_handler.bal:
```diff
@@ -112,50 +112,50 @@ function getApplicationStatusByApplicationId(string applicationId) returns strin
+# add a new version of an application with the incoming form details.
 # 
 # + form - Form containing the tree removal data.
+# + applicationId - The Id of the application which should be altered.
```
```suggestion
# + applicationId - The Id of the application which should be altered or versioned.
```

deshankoswatte commented on core/src/env_platform/db_handler.bal:
```diff
@@ -112,50 +112,50 @@ function getApplicationStatusByApplicationId(string applicationId) returns strin
 # 
 # + form - Form containing the tree removal data.
+# + applicationId - The Id of the application which should be altered.
 # + return - This function will return true if draft is updated in the database, false if not or else it returns a mongodb:Database error.
```
```suggestion
# + return - This function will return true if draft is updated or the application is versioned in the database, 
# false if not or else it returns a mongodb:Database error.
```

deshankoswatte commented on core/src/env_platform/utility.bal:
```diff
@@ -48,3 +48,39 @@ function extractTreeInformationAsJSONArray(TreeInformation[] treeInfoArray) retu
+# 
+# + form - TreeRemovalForm type record.
+# + return - Returns the map<json> which suites for the database
+function constructApplicationVersion(TreeRemovalForm form) returns map<json> {
```
Will the name `constructApplicationVersion` be suitable since we use this to perform altering if application is a draft and versioning if application is already submitted? @sanjulamadurapperuma @yashodgayashan thoughts?

yashodgayashan replied:
Should it be constructApplication or?

sanjulamadurapperuma replied:
Yes I missed this, it has to be generic

deshankoswatte replied:
Yeah it should be general like `constructApplication`.

yashodgayashan replied:
Any suggestions?

deshankoswatte replied:
`constructApplication` is good enough i guess.

yashodgayashan replied:
Yes. It is

deshankoswatte commented on core/src/env_platform/utility.bal:
```diff
@@ -48,3 +48,39 @@ function extractTreeInformationAsJSONArray(TreeInformation[] treeInfoArray) retu
+# The `constructApplicationVersion` function will construct the application version which will be suitable for the database.
+# 
+# + form - TreeRemovalForm type record.
+# + return - Returns the map<json> which suites for the database
```
```suggestion
# + return - Returns a map<json> containing the application structure which suites the database.
```

deshankoswatte commented on core/src/env_platform/utility.bal:
```diff
@@ -48,3 +48,39 @@ function extractTreeInformationAsJSONArray(TreeInformation[] treeInfoArray) retu
+
+# The `constructApplicationVersion` function will construct the application version which will be suitable for the database.
+# 
+# + form - TreeRemovalForm type record.
```
```suggestion
# + form - Form containing the tree removal data.
```

deshankoswatte commented on core/src/env_platform/utility.bal:
```diff
@@ -48,3 +48,39 @@ function extractTreeInformationAsJSONArray(TreeInformation[] treeInfoArray) retu
     return treeInformation;
 }
+
+# The `constructApplicationVersion` function will construct the application version which will be suitable for the database.
```
Will have to change this based on the validity of the question about the method name suitable.

sanjulamadurapperuma replied:
```suggestion
# The `constructApplication` function will construct the application which will be suitable for the database.
```

sanjulamadurapperuma commented on core/src/env_platform/utility.bal:
```diff
@@ -48,3 +48,39 @@ function extractTreeInformationAsJSONArray(TreeInformation[] treeInfoArray) retu
+# 
+# + form - Form containing the tree removal data.
+# + return - Returns a map<json> containing the application structure which suites the database.
+function constructApplicationVersion(TreeRemovalForm form) returns map<json> {
```
```suggestion
function constructApplication(TreeRemovalForm form) returns map<json> {
```

yashodgayashan commented:
@sanjulamadurapperuma @deshankoswatte Resolved requested changes
